File: Oreon/Oreon_Java/oe-vk-components/src/main/java/org/oreon/vk/components/atmosphere/Atmosphere.java
package org.oreon.vk.components.atmosphere;

import org.lwjgl.vulkan.VkPhysicalDeviceMemoryProperties;
import org.oreon.core.context.BaseContext;
import org.oreon.core.model.Mesh;
import org.oreon.core.model.Vertex.VertexLayout;
import org.oreon.core.scenegraph.NodeComponentType;
import org.oreon.core.scenegraph.Renderable;
import org.oreon.core.util.BufferUtil;
import org.oreon.core.util.Constants;
import org.oreon.core.util.ProceduralTexturing;
import org.oreon.core.vk.command.CommandBuffer;
import org.oreon.core.vk.context.DeviceManager.DeviceType;
import org.oreon.core.vk.context.VkContext;
import org.oreon.core.vk.context.VkResources.VkDescriptorName;
import org.oreon.core.vk.descriptor.DescriptorSet;
import org.oreon.core.vk.descriptor.DescriptorSetLayout;
import org.oreon.core.vk.device.LogicalDevice;
import org.oreon.core.vk.memory.VkBuffer;
import org.oreon.core.vk.pipeline.ShaderModule;
import org.oreon.core.vk.pipeline.ShaderPipeline;
import org.oreon.core.vk.pipeline.VkPipeline;
import org.oreon.core.vk.pipeline.VkVertexInput;
import org.oreon.core.vk.scenegraph.VkMeshData;
import org.oreon.core.vk.scenegraph.VkRenderInfo;
import org.oreon.core.vk.util.VkAssimpModelLoader;
import org.oreon.core.vk.util.VkUtil;
import org.oreon.core.vk.wrapper.buffer.VkBufferHelper;
import org.oreon.core.vk.wrapper.buffer.VkUniformBuffer;
import org.oreon.core.vk.wrapper.command.SecondaryDrawIndexedCmdBuffer;
import org.oreon.core.vk.wrapper.pipeline.GraphicsPipeline;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

import static org.lwjgl.system.MemoryUtil.memAlloc;
import static org.lwjgl.vulkan.VK10.*;

public class Atmosphere extends Renderable {

    private final VkUniformBuffer uniformBuffer;

    public Atmosphere() {

        LogicalDevice device = VkContext.getDeviceManager().getLogicalDevice(DeviceType.MAJOR_GRAPHICS_DEVICE);
        VkPhysicalDeviceMemoryProperties memoryProperties =
                VkContext.getDeviceManager().getPhysicalDevice(DeviceType.MAJOR_GRAPHICS_DEVICE).getMemoryProperties();

        getWorldTransform().setLocalScaling(Constants.ZFAR * 0.5f, Constants.ZFAR * 0.5f, Constants.ZFAR * 0.5f);

        Mesh mesh = VkAssimpModelLoader.loadModel("models/obj/dome", "dome.obj").get(0).getMesh();
        ProceduralTexturing.dome(mesh);

        ByteBuffer ubo = memAlloc(Float.BYTES * 16);
        ubo.put(BufferUtil.createByteBuffer(getWorldTransform().getWorldMatrix()));
        ubo.flip();

        uniformBuffer = new VkUniformBuffer(device.getHandle(), memoryProperties, ubo);

        ShaderModule vertexShader = new ShaderModule(device.getHandle(),
                "shaders/atmosphere/atmosphere.vert.spv", VK_SHADER_STAGE_VERTEX_BIT);

        ShaderPipeline graphicsShaderPipeline = new ShaderPipeline(device.getHandle());
        graphicsShaderPipeline.addShaderModule(vertexShader);
        graphicsShaderPipeline.createFragmentShader(BaseContext.getConfig().isAtmosphericScatteringEnable() ?
                "shaders/atmosphere/atmospheric_scattering.frag.spv" : "shaders/atmosphere/atmosphere.frag.spv");
        graphicsShaderPipeline.createShaderPipeline();

        ShaderPipeline reflectionShaderPipeline = new ShaderPipeline(device.getHandle());
        reflectionShaderPipeline.addShaderModule(vertexShader);
        reflectionShaderPipeline.createFragmentShader("shaders/atmosphere/atmospheric_scattering_reflection.frag.spv");
        reflectionShaderPipeline.createShaderPipeline();

        List<DescriptorSet> descriptorSets = new ArrayList<DescriptorSet>();
        List<DescriptorSetLayout> descriptorSetLayouts = new ArrayList<DescriptorSetLayout>();

        DescriptorSetLayout descriptorSetLayout = new DescriptorSetLayout(device.getHandle(), 1);
        descriptorSetLayout.addLayoutBinding(0, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,
                VK_SHADER_STAGE_VERTEX_BIT);
        descriptorSetLayout.create();
        DescriptorSet descriptorSet = new DescriptorSet(device.getHandle(),
                device.getDescriptorPool(Thread.currentThread().getId()).getHandle(),
                descriptorSetLayout.getHandlePointer());
        descriptorSet.updateDescriptorBuffer(uniformBuffer.getHandle(),
                ubo.limit(), 0, 0, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER);

        descriptorSets.add(VkContext.getCamera().getDescriptorSet());
        descriptorSets.add(descriptorSet);
        descriptorSets.add(VkContext.getResources().getDescriptors().get(VkDescriptorName.DIRECTIONAL_LIGHT).getDescriptorSet());
        descriptorSetLayouts.add(VkContext.getCamera().getDescriptorSetLayout());
        descriptorSetLayouts.add(descriptorSetLayout);
        descriptorSetLayouts.add(VkContext.getResources().getDescriptors().get(VkDescriptorName.DIRECTIONAL_LIGHT).getDescriptorSetLayout());

        VkVertexInput vertexInput = new VkVertexInput(VertexLayout.POS);

        ByteBuffer vertexBuffer = BufferUtil.createByteBuffer(mesh.getVertices(), VertexLayout.POS);
        ByteBuffer indexBuffer = BufferUtil.createByteBuffer(mesh.getIndices());

        int pushConstantsRange = Float.BYTES * 20 + Integer.BYTES * 3;

        ByteBuffer pushConstants = memAlloc(pushConstantsRange);
        pushConstants.put(BufferUtil.createByteBuffer(VkContext.getCamera().getProjectionMatrix()));
        pushConstants.putFloat(BaseContext.getConfig().getSunRadius());
        pushConstants.putInt(BaseContext.getConfig().getFrameWidth());
        pushConstants.putInt(BaseContext.getConfig().getFrameHeight());
        pushConstants.putInt(0);
        pushConstants.putFloat(BaseContext.getConfig().getAtmosphereBloomFactor());
        pushConstants.putFloat(BaseContext.getConfig().getHorizonVerticalShift());
        pushConstants.putFloat(BaseContext.getConfig().getHorizonReflectionVerticalShift());
        pushConstants.flip();

        VkPipeline graphicsPipeline = new GraphicsPipeline(device.getHandle(),
                graphicsShaderPipeline, vertexInput, VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST,
                VkUtil.createLongBuffer(descriptorSetLayouts),
                BaseContext.getConfig().getFrameWidth(),
                BaseContext.getConfig().getFrameHeight(),
                VkContext.getResources().getOffScreenFbo().getRenderPass().getHandle(),
                VkContext.getResources().getOffScreenFbo().getColorAttachmentCount(),
                BaseContext.getConfig().getMultisampling_sampleCount(),
                pushConstantsRange, VK_SHADER_STAGE_FRAGMENT_BIT);

        VkBuffer vertexBufferObject = VkBufferHelper.createDeviceLocalBuffer(
                device.getHandle(), memoryProperties,
                device.getTransferCommandPool(Thread.currentThread().getId()).getHandle(),
                device.getTransferQueue(),
                vertexBuffer, VK_BUFFER_USAGE_VERTEX_BUFFER_BIT);

        VkBuffer indexBufferObject = VkBufferHelper.createDeviceLocalBuffer(
                device.getHandle(), memoryProperties,
                device.getTransferCommandPool(Thread.currentThread().getId()).getHandle(),
                device.getTransferQueue(),
                indexBuffer, VK_BUFFER_USAGE_INDEX_BUFFER_BIT);

        CommandBuffer mainCommandBuffer = new SecondaryDrawIndexedCmdBuffer(
                device.getHandle(),
                device.getGraphicsCommandPool(Thread.currentThread().getId()).getHandle(),
                graphicsPipeline.getHandle(), graphicsPipeline.getLayoutHandle(),
                VkContext.getResources().getOffScreenFbo().getFrameBuffer().getHandle(),
                VkContext.getResources().getOffScreenFbo().getRenderPass().getHandle(),
                0,
                VkUtil.createLongArray(descriptorSets),
                vertexBufferObject.getHandle(),
                indexBufferObject.getHandle(),
                mesh.getIndices().length,
                pushConstants, VK_SHADER_STAGE_FRAGMENT_BIT);

        VkMeshData meshData = VkMeshData.builder().vertexBufferObject(vertexBufferObject)
                .vertexBuffer(vertexBuffer).indexBufferObject(indexBufferObject).indexBuffer(indexBuffer)
                .build();
        VkRenderInfo mainRenderInfo = VkRenderInfo.builder().commandBuffer(mainCommandBuffer)
                .pipeline(graphicsPipeline).descriptorSets(descriptorSets)
                .descriptorSetLayouts(descriptorSetLayouts).build();


        addComponent(NodeComponentType.MESH_DATA, meshData);
        addComponent(NodeComponentType.MAIN_RENDERINFO, mainRenderInfo);
        addComponent(NodeComponentType.WIREFRAME_RENDERINFO, mainRenderInfo);

        if (VkContext.getResources().getReflectionFbo() != null) {
            VkPipeline reflectionPipeline = new GraphicsPipeline(device.getHandle(),
                    reflectionShaderPipeline, vertexInput, VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST,
                    VkUtil.createLongBuffer(descriptorSetLayouts),
                    VkContext.getResources().getReflectionFbo().getWidth(),
                    VkContext.getResources().getReflectionFbo().getHeight(),
                    VkContext.getResources().getReflectionFbo().getRenderPass().getHandle(),
                    VkContext.getResources().getReflectionFbo().getColorAttachmentCount(), 1,
                    pushConstantsRange, VK_SHADER_STAGE_FRAGMENT_BIT);

            CommandBuffer reflectionCommandBuffer = new SecondaryDrawIndexedCmdBuffer(
                    device.getHandle(),
                    device.getGraphicsCommandPool(Thread.currentThread().getId()).getHandle(),
                    reflectionPipeline.getHandle(), reflectionPipeline.getLayoutHandle(),
                    VkContext.getResources().getReflectionFbo().getFrameBuffer().getHandle(),
                    VkContext.getResources().getReflectionFbo().getRenderPass().getHandle(),
                    0,
                    VkUtil.createLongArray(descriptorSets),
                    vertexBufferObject.getHandle(),
                    indexBufferObject.getHandle(),
                    mesh.getIndices().length,
                    pushConstants, VK_SHADER_STAGE_FRAGMENT_BIT);

            VkRenderInfo reflectionRenderInfo = VkRenderInfo.builder().commandBuffer(reflectionCommandBuffer)
                    .pipeline(reflectionPipeline).build();

            addComponent(NodeComponentType.REFLECTION_RENDERINFO, reflectionRenderInfo);
        }

        graphicsShaderPipeline.destroy();
        reflectionShaderPipeline.destroy();
    }

    public void update() {
        super.update();

        uniformBuffer.mapMemory(BufferUtil.createByteBuffer(getWorldTransform().getWorldMatrix()));
    }

    public void shutdown() {

        super.shutdown();
        uniformBuffer.destroy();
    }

}

File: Oreon/Oreon_Kotlin/src/oe-vk-components/src/main/java/org/oreon/vk/components/atmosphere/Atmosphere.kt
package org.oreon.vk.components.atmosphere

import org.lwjgl.system.MemoryUtil
import org.lwjgl.vulkan.VK10
import org.oreon.core.context.BaseContext
import org.oreon.core.model.Vertex.VertexLayout
import org.oreon.core.scenegraph.NodeComponentType
import org.oreon.core.scenegraph.Renderable
import org.oreon.core.util.BufferUtil.createByteBuffer
import org.oreon.core.util.Constants
import org.oreon.core.util.ProceduralTexturing.dome
import org.oreon.core.vk.command.CommandBuffer
import org.oreon.core.vk.context.DeviceManager.DeviceType
import org.oreon.core.vk.context.VkContext.deviceManager
import org.oreon.core.vk.context.VkContext.getCamera
import org.oreon.core.vk.context.VkContext.resources
import org.oreon.core.vk.context.VkResources.VkDescriptorName
import org.oreon.core.vk.descriptor.DescriptorSet
import org.oreon.core.vk.descriptor.DescriptorSetLayout
import org.oreon.core.vk.pipeline.ShaderModule
import org.oreon.core.vk.pipeline.ShaderPipeline
import org.oreon.core.vk.pipeline.VkPipeline
import org.oreon.core.vk.pipeline.VkVertexInput
import org.oreon.core.vk.scenegraph.VkMeshData
import org.oreon.core.vk.scenegraph.VkRenderInfo
import org.oreon.core.vk.util.VkAssimpModelLoader.loadModel
import org.oreon.core.vk.util.VkUtil.createLongArray
import org.oreon.core.vk.util.VkUtil.createLongBuffer
import org.oreon.core.vk.wrapper.buffer.VkBufferHelper.createDeviceLocalBuffer
import org.oreon.core.vk.wrapper.buffer.VkUniformBuffer
import org.oreon.core.vk.wrapper.command.SecondaryDrawIndexedCmdBuffer
import org.oreon.core.vk.wrapper.pipeline.GraphicsPipeline
import java.util.*

class Atmosphere : Renderable() {
    private val uniformBuffer: VkUniformBuffer
    override fun update() {
        super.update()
        uniformBuffer.mapMemory(createByteBuffer(worldTransform.worldMatrix))
    }

    override fun shutdown() {
        super.shutdown()
        uniformBuffer.destroy()
    }

    init {
        val device = deviceManager.getLogicalDevice(DeviceType.MAJOR_GRAPHICS_DEVICE)
        val memoryProperties = deviceManager.getPhysicalDevice(DeviceType.MAJOR_GRAPHICS_DEVICE).memoryProperties
        worldTransform.setLocalScaling(Constants.ZFAR * 0.5f, Constants.ZFAR * 0.5f, Constants.ZFAR * 0.5f)
        val mesh = loadModel("models/obj/dome", "dome.obj")[0].mesh
        dome(mesh!!)
        val ubo = MemoryUtil.memAlloc(java.lang.Float.BYTES * 16)
        ubo.put(createByteBuffer(worldTransform.worldMatrix))
        ubo.flip()
        uniformBuffer = VkUniformBuffer(device.handle, memoryProperties, ubo)
        val vertexShader = ShaderModule(device.handle,
                "shaders/atmosphere/atmosphere.vert.spv", VK10.VK_SHADER_STAGE_VERTEX_BIT)
        val graphicsShaderPipeline = ShaderPipeline(device.handle)
        graphicsShaderPipeline.addShaderModule(vertexShader)
        graphicsShaderPipeline.createFragmentShader(if (BaseContext.config.AtmosphericScatteringEnable) "shaders/atmosphere/atmospheric_scattering.frag.spv" else "shaders/atmosphere/atmosphere.frag.spv")
        graphicsShaderPipeline.createShaderPipeline()
        val reflectionShaderPipeline = ShaderPipeline(device.handle)
        reflectionShaderPipeline.addShaderModule(vertexShader)
        reflectionShaderPipeline.createFragmentShader("shaders/atmosphere/atmospheric_scattering_reflection.frag.spv")
        reflectionShaderPipeline.createShaderPipeline()
        val descriptorSets: MutableList<DescriptorSet> = ArrayList()
        val descriptorSetLayouts: MutableList<DescriptorSetLayout> = ArrayList()
        val descriptorSetLayout = DescriptorSetLayout(device.handle, 1)
        descriptorSetLayout.addLayoutBinding(0, VK10.VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,
                VK10.VK_SHADER_STAGE_VERTEX_BIT)
        descriptorSetLayout.create()
        val descriptorSet = DescriptorSet(device.handle,
                device.getDescriptorPool(Thread.currentThread().id)!!.handle,
                descriptorSetLayout.handlePointer)
        descriptorSet.updateDescriptorBuffer(uniformBuffer.handle,
                ubo.limit().toLong(), 0, 0, VK10.VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER)
        descriptorSets.add(getCamera().descriptorSet)
        descriptorSets.add(descriptorSet)
        descriptorSets.add(resources.descriptors[VkDescriptorName.DIRECTIONAL_LIGHT]!!.descriptorSet)
        descriptorSetLayouts.add(getCamera().descriptorSetLayout)
        descriptorSetLayouts.add(descriptorSetLayout)
        descriptorSetLayouts.add(resources.descriptors[VkDescriptorName.DIRECTIONAL_LIGHT]!!.descriptorSetLayout)
        val vertexInput = VkVertexInput(VertexLayout.POS)
        val vertexBuffer = createByteBuffer(mesh.vertices, VertexLayout.POS)
        val indexBuffer = createByteBuffer(*mesh.indices)
        val pushConstantsRange = java.lang.Float.BYTES * 20 + Integer.BYTES * 3
        val pushConstants = MemoryUtil.memAlloc(pushConstantsRange)
        pushConstants.put(createByteBuffer(getCamera().projectionMatrix))
        pushConstants.putFloat(BaseContext.config.sunRadius)
        pushConstants.putInt(BaseContext.config.frameWidth)
        pushConstants.putInt(BaseContext.config.frameHeight)
        pushConstants.putInt(0)
        pushConstants.putFloat(BaseContext.config.atmosphereBloomFactor)
        pushConstants.putFloat(BaseContext.config.horizonVerticalShift)
        pushConstants.putFloat(BaseContext.config.horizonVerticalShift * 2)
        pushConstants.flip()
        val graphicsPipeline: VkPipeline = GraphicsPipeline(device.handle,
                graphicsShaderPipeline, vertexInput, VK10.VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST,
                createLongBuffer(descriptorSetLayouts),
                BaseContext.config.frameWidth,
                BaseContext.config.frameHeight,
                resources.offScreenFbo!!.renderPass!!.handle,
                resources.offScreenFbo!!.colorAttachmentCount,
                BaseContext.config.multisampling_sampleCount,
                pushConstantsRange, VK10.VK_SHADER_STAGE_FRAGMENT_BIT)
        val vertexBufferObject = createDeviceLocalBuffer(
                device.handle, memoryProperties,
                device.getTransferCommandPool(Thread.currentThread().id)!!.handle,
                device.transferQueue,
                vertexBuffer, VK10.VK_BUFFER_USAGE_VERTEX_BUFFER_BIT)
        val indexBufferObject = createDeviceLocalBuffer(
                device.handle, memoryProperties,
                device.getTransferCommandPool(Thread.currentThread().id)!!.handle,
                device.transferQueue,
                indexBuffer, VK10.VK_BUFFER_USAGE_INDEX_BUFFER_BIT)
        val mainCommandBuffer: CommandBuffer = SecondaryDrawIndexedCmdBuffer(
                device.handle,
                device.getGraphicsCommandPool(Thread.currentThread().id)!!.handle,
                graphicsPipeline.handle, graphicsPipeline.layoutHandle,
                resources.offScreenFbo!!.frameBuffer!!.handle,
                resources.offScreenFbo!!.renderPass!!.handle,
                0,
                createLongArray(descriptorSets),
                vertexBufferObject.handle,
                indexBufferObject.handle,
                mesh.indices.size,
                pushConstants, VK10.VK_SHADER_STAGE_FRAGMENT_BIT)
        val meshData: VkMeshData = VkMeshData(vertexBufferObject = vertexBufferObject,
                vertexBuffer = vertexBuffer, indexBufferObject = indexBufferObject, indexBuffer = indexBuffer)
        val mainRenderInfo: VkRenderInfo = VkRenderInfo(commandBuffer = mainCommandBuffer,
                pipeline = graphicsPipeline, descriptorSets = descriptorSets,
                descriptorSetLayouts = descriptorSetLayouts)
        addComponent(NodeComponentType.MESH_DATA, meshData)
        addComponent(NodeComponentType.MAIN_RENDERINFO, mainRenderInfo)
        addComponent(NodeComponentType.WIREFRAME_RENDERINFO, mainRenderInfo)
        if (resources.reflectionFbo != null) {
            val reflectionPipeline: VkPipeline = GraphicsPipeline(device.handle,
                    reflectionShaderPipeline, vertexInput, VK10.VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST,
                    createLongBuffer(descriptorSetLayouts),
                    resources.reflectionFbo!!.width,
                    resources.reflectionFbo!!.height,
                    resources.reflectionFbo!!.renderPass!!.handle,
                    resources.reflectionFbo!!.colorAttachmentCount, 1,
                    pushConstantsRange, VK10.VK_SHADER_STAGE_FRAGMENT_BIT)
            val reflectionCommandBuffer: CommandBuffer = SecondaryDrawIndexedCmdBuffer(
                    device.handle,
                    device.getGraphicsCommandPool(Thread.currentThread().id)!!.handle,
                    reflectionPipeline.handle, reflectionPipeline.layoutHandle,
                    resources.reflectionFbo!!.frameBuffer!!.handle,
                    resources.reflectionFbo!!.renderPass!!.handle,
                    0,
                    createLongArray(descriptorSets),
                    vertexBufferObject.handle,
                    indexBufferObject.handle,
                    mesh.indices.size,
                    pushConstants, VK10.VK_SHADER_STAGE_FRAGMENT_BIT)
            val reflectionRenderInfo: VkRenderInfo = VkRenderInfo(commandBuffer = reflectionCommandBuffer,
                    pipeline = reflectionPipeline)
            addComponent(NodeComponentType.REFLECTION_RENDERINFO, reflectionRenderInfo)
        }
        graphicsShaderPipeline.destroy()
        reflectionShaderPipeline.destroy()
    }
}